src/files/DCDFile.cpp

```diff
@@ -0,0 +1,457 @@
+      fileStream.read((char*)freeAtoms, sizeof(int) * numFreeAtoms);
+      fileStream.read((char*)&fortcheck2, sizeof(unsigned int));
+      if (!checkFortranIOerror(__FILE__, __LINE__, fortcheck1, fortcheck2))
+      {
+         return;
+      }
+   }
+
+   headByteNum = std::abs(initialPos - fileStream.tellg());
+}
+
+void DCDFile::writeHeader()
+{
+   if (fileStream.is_open())
+   {
+      unsigned int icntrlBlockSize = (sizeof(char) * 4) + (sizeof(int) * 20);
+      fileStream.write((char*)&icntrlBlockSize, sizeof(unsigned int));
+      fileStream.write((char*)&HDR, sizeof(char) * 4);                //first data block written by fortran  : a character array of length 4.
+      fileStream.write((char*)ICNTRL, sizeof(int) * 20);                    //second data block written by fortran : an integer(4) array of length 20.
+      fileStream.write((char*)&icntrlBlockSize, sizeof(unsigned int));
+
+
+      /* Several "lines" of title of length 80 are written to the dcd file by CHARMM */
+      unsigned int titlesBlockSize = sizeof(int) + (NTITLE * 80) + 1;
+      fileStream.write((char*)&titlesBlockSize, sizeof(unsigned int));
+      fileStream.write((char*)&NTITLE, sizeof(int));
+      for (int it = 0; it < NTITLE; it++)
+      {
+         fileStream.write((char*)&TITLE[it * 80], sizeof(char) * 80);
+      }
+      TITLE[NTITLE * 80] = '\0';
+      fileStream.write((char*)&titlesBlockSize, sizeof(unsigned int));
+
+      // writing number of atoms
+      unsigned int atomsNumBlockSize = numAtoms * sizeof(int);
+      fileStream.write((char*)&atomsNumBlockSize, sizeof(unsigned int));
+      fileStream.write((char*)&numAtoms, sizeof(int));
+      fileStream.write((char*)&atomsNumBlockSize, sizeof(unsigned int));
+
+      /* If some atoms of the MD or MC simulation are frozen (i.e. never moving ) it is useless to store their coordinates more than once.
+       * In that case a list of Free atoms (moving ones) is written at the end of the header part of the dcd.
+       */
+      if (numFreeAtoms != numAtoms)
+      {
+         unsigned int freeAtomsBlockSize = numFreeAtoms * sizeof(int);
+         fileStream.write((char*)&freeAtomsBlockSize, sizeof(unsigned int));
+         fileStream.write((char*)freeAtoms, sizeof(int) * numFreeAtoms);
+         fileStream.write((char*)&freeAtomsBlockSize, sizeof(unsigned int));
+      }
+   }
+}
+
+void DCDFile::readFrame(std::vector<float>& X, std::vector<float>& Y,
+   std::vector<float>& Z, int toFrame)
+{
+   //dcd files store all the atoms in the first frame (even if there are some frozen atoms)
+   //and only store free atoms in the other frames
+   int frameAtmNum;
+   if (ICNTRL[8] != 0 && frameNumber != 0)
+   {
+      frameAtmNum = numAtoms - ICNTRL[8];
+   }
+   else
+   {
+      frameAtmNum = numAtoms;
+   }
+
+   X.reserve(frameAtmNum);
+   Y.reserve(frameAtmNum);
+   Z.reserve(frameAtmNum);
+
+   unsigned int fortcheck1, fortcheck2;
+
+   float* tmpX = new float[frameAtmNum];
```

Comment:
why allocate extra memory? This could be replaced by `float* tmpX = X.data()`, and the copy below removed